x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java

```diff
@@ -470,6 +473,15 @@ protected PhysicalPlan rule(AggregateExec aggregate) {
                         );
                     }
                 }
+                if (exec instanceof FilterExec filterExec) {
+                    // Note that ST_CENTROID does not support shapes, but SpatialRelatesFunction does, so when we extend the centroid
+                    // to support shapes, we need to consider loading shape doc-values for both centroid and relates (ST_INTERSECTS)
+                    if (filterExec.condition() instanceof SpatialRelatesFunction spatialRelatesFunction) {
```

Comment:
Does it matter here if the condition is a boolean composed of multiple other conditions and the spatial function is just one of those conditions?

Reply:
This challenge here is that if we optimized ST_CENTROID for reading points as doc-values, then any spatial relates function also needs to expect doc-values, otherwise it fails to interpret the types, since spatial doc-values are completely different to spatial types from source. So this code is making sure that both ST_CENTROID and ST_INTERSECTS know that the data will arrive in a different format.

Reply:
What I'm after here (and failed to double check this with an actual query) is the type of the filter you are trying to check. Can `filterExec.condition()` be something like `age > 0 and gender='F' and st_centroid......` ? If so, should you look for a `SpatialRelatesFunction` in the filter and not only check that the filter itself is of type `SpatialRelatesFunction`?

In short

`age > 0 and gender='F' and st_centroid......`
vs
`st_centroid......`

Reply:
Right, I see what you mean. I misunderstood your original comment. I will investigate this, and suspect you are right and I'm not dealing with compound boolean expressions correctly.

Reply:
I pushed a fix for this. Curiously it only matters for CsvTests, because the lucene-pushdown of ST_INTERSECTS actually erases this scenario.